includes/class-sensei-analysis-overview-list-table.php

```diff
@@ -369,18 +372,53 @@ protected function get_row_data( $item ) {
 					$course_title = '<strong><a class="row-title" href="' . esc_url( $url ) . '">' . apply_filters( 'the_title', $item->post_title, $item->ID ) . '</a></strong>';
 				}
 
-				$column_data = apply_filters(
-					'sensei_analysis_overview_column_data',
-					array(
-						'title'              => $course_title,
-						'last_activity'      => $last_activity_date,
-						'completions'        => $course_completions,
-						'average_percent'    => $average_grade,
-						'days_to_completion' => $average_completion_days,
-					),
-					$item,
-					$this
+				// Fetch learners in course.
+				$course_args           = array(
+					'post_id' => $item->ID,
+					'type'    => 'sensei_course_status',
+					'status'  => array( 'in-progress', 'complete' ),
+				);
+				$course_students_count = Sensei_Utils::sensei_check_for_activity( $course_args );
+
+				// Calculate average lesson progress per student for course.
+				$average_course_progress = 0;
+
+				// Get all course lessons.
+				$lessons        = Sensei()->course->course_lessons( $item->ID, 'publish', 'ids' );
+				$course_lessons = is_array( $lessons ) ? $lessons : array( $lessons );
+				$total_lessons  = count( $course_lessons );
+
+				// Get all completed lessons.
+				$lesson_args     = array(
+					'post__in' => $course_lessons,
+					'type'     => 'sensei_lesson_status',
+					'status'   => array( 'graded', 'ungraded', 'passed', 'failed', 'complete' ),
+					'count'    => true,
 				);
+				$completed_count = (int) Sensei_Utils::sensei_check_for_activity( apply_filters( 'sensei_analysis_course_lesson_completions', $lesson_args, $item ) );
+				// Calculate average progress.
+				$average_course_progress = __( 'N/A', 'sensei-lms' );
+				if ( $course_students_count && $total_lessons ) {
+					// Average course progress is calculated based on lessons completed for the course
+					// divided by the total possible lessons completed.
+					$average_course_progress_value = $completed_count / ( $course_students_count * $total_lessons ) * 100;
+					$average_course_progress       = esc_html(
```

Comment:
For me, escaping here looks redundant as we completely control the result.

Reply:
It's generally considered a best practice to always escape, and to do so as late as possible (i.e. at the time that the value is being displayed, so technically in this case that would mean in the `$column_data` assignment), even if we control the value. The reason for this is so that if a value should change to an untrusted value in the future, and we haven't escaped it at the time of display, we'll have created a security issue.